src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts

```diff
@@ -0,0 +1,517 @@
+      if (
+        !(await this.validateWorkspacesAndSavedObjectsPermissions(
+          objectToDeleted,
+          wrapperOptions.request,
+          [WorkspacePermissionMode.LibraryWrite],
+          [WorkspacePermissionMode.Write]
+        ))
+      ) {
+        throw generateSavedObjectsPermissionError();
+      }
+      return await wrapperOptions.client.delete(type, id, options);
+    };
+
+    // validate `objectToUpdate` if can update with workspace permission, which is used for update and bulkUpdate
+    const validateUpdateWithWorkspacePermission = async <T = unknown>(
+      objectToUpdate: SavedObject<T>
+    ): Promise<boolean> => {
+      return await this.validateWorkspacesAndSavedObjectsPermissions(
+        objectToUpdate,
+        wrapperOptions.request,
+        [WorkspacePermissionMode.LibraryWrite],
+        [WorkspacePermissionMode.Write],
+        false
+      );
+    };
+
+    const updateWithWorkspacePermissionControl = async <T = unknown>(
+      type: string,
+      id: string,
+      attributes: Partial<T>,
+      options: SavedObjectsUpdateOptions = {}
+    ): Promise<SavedObjectsUpdateResponse<T>> => {
+      const objectToUpdate = await wrapperOptions.client.get<T>(type, id, options);
+      const permitted = await validateUpdateWithWorkspacePermission(objectToUpdate);
+      if (!permitted) {
+        throw generateSavedObjectsPermissionError();
+      }
+      return await wrapperOptions.client.update(type, id, attributes, options);
+    };
+
+    const bulkUpdateWithWorkspacePermissionControl = async <T = unknown>(
+      objects: Array<SavedObjectsBulkUpdateObject<T>>,
+      options?: SavedObjectsBulkUpdateOptions
+    ): Promise<SavedObjectsBulkUpdateResponse<T>> => {
+      const objectsToUpdate = await wrapperOptions.client.bulkGet<T>(objects, options);
+
+      for (const object of objectsToUpdate.saved_objects) {
+        const permitted = await validateUpdateWithWorkspacePermission(object);
+        if (!permitted) {
+          throw generateSavedObjectsPermissionError();
+        }
+      }
+
+      return await wrapperOptions.client.bulkUpdate(objects, options);
+    };
+
+    const bulkCreateWithWorkspacePermissionControl = async <T = unknown>(
+      objects: Array<SavedObjectsBulkCreateObject<T>>,
+      options: SavedObjectsCreateOptions = {}
+    ): Promise<SavedObjectsBulkResponse<T>> => {
+      const hasTargetWorkspaces = options?.workspaces && options.workspaces.length > 0;
+
+      if (
+        hasTargetWorkspaces &&
+        !(await this.validateMultiWorkspacesPermissions(
+          options.workspaces ?? [],
+          wrapperOptions.request,
+          [WorkspacePermissionMode.LibraryWrite]
+        ))
+      ) {
+        throw generateSavedObjectsPermissionError();
+      }
+
+      /**
+       *
+       * If target workspaces parameter exists, we don't need to do permission validation again.
+       * The bulk create method in repository doesn't allow extends workspaces with override.
```

Comment:
The workspace conflict validation logic is under the repository `bulkCreate` method before. Due to #254 , these logic has been moved to the workspace conflict validation saved objects wrapper. The `bulkCreate` method can't override existing saved objects with extra workspaces. For example, object-1 is in workspace-a. It can't be `bulkCreate` to workspace-a and workspace-b.  The workspace-a and workspace-b is `options.workspaces`, and it should be a subset of object-1's workspace. We already validate the `options.workspaces` in begin.  So we don't need validate the permissions of `options.workspaces` anymore.